### PR TITLE
Improve `Amount` and `Currency`

### PR DESCRIPTION
**Pull request by LimpidCrypto**
### Model conversion
This PR changes the lifetime of `Currency` and `Amount` from `'static` to `'a`. It also adds `to_amount` to turn a `Currency` into an `Amount` giving a `value`. Implement the  `From<Amount> for Currency` trait to turn an `Amount` into a `Currency`.

### `value` conversion
The `get_value_as_u32` method was also replaced with `TryInto<rust_decimal::Decimal> for Amount` to have a more convenient way to deal with an `Amount.value`. The `get_value_as_u32` also don't make sense for `IssuedCurrencyAmount` since it's `value` is most likely a decimal value.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Future Tasks
- [ ] Add errors as soon as #54 was merged.

### CODE REVIEW

File: src/models/transactions/nftoken_accept_offer.rs [hunk 1]
```diff
@@ -141,11 +143,14 @@ impl<'a> NFTokenAcceptOfferError for NFTokenAcceptOffer<'a> {
...
     }
     fn _get_nftoken_broker_fee_error(&self) -> Result<(), NFTokenAcceptOfferException> {
-        match self.nftoken_broker_fee.as_ref() {
-            Some(nftoken_broker_fee) => match nftoken_broker_fee.get_value_as_u32() == 0 {
-                true => Err(NFTokenAcceptOfferException::InvalidBrokerFeeMustBeGreaterZero),
-                false => Ok(()),
-            },
+        match &self.nftoken_broker_fee {
+            Some(nftoken_broker_fee) => {
+                let as_decimal: Decimal = nftoken_broker_fee.clone().try_into().unwrap();
```
Comment (sephynox) [inline src/models/transactions/nftoken_accept_offer.rs:148]: This is an unhandled `try_into` and `unwrap`.

File: src/models/transactions/nftoken_accept_offer.rs [hunk 2]
```diff
@@ -141,11 +143,14 @@ impl<'a> NFTokenAcceptOfferError for NFTokenAcceptOffer<'a> {
...
+                match as_decimal.is_zero() {
+                    true => Err(NFTokenAcceptOfferException::InvalidBrokerFeeMustBeGreaterZero),
+                    false => Ok(()),
+                }
```
Comment (sephynox) [inline src/models/transactions/nftoken_accept_offer.rs:152]: We can just use if/else blocks for booleans.

File: src/models/transactions/nftoken_create_offer.rs [hunk 3]
```diff
@@ -181,9 +183,10 @@ impl<'a> Transaction for NFTokenCreateOffer<'a> {
...
 impl<'a> NFTokenCreateOfferError for NFTokenCreateOffer<'a> {
     fn _get_amount_error(&self) -> Result<(), NFTokenCreateOfferException> {
+        let amount_as_decimal: Decimal = self.amount.clone().try_into().unwrap();
         match !self.has_flag(&Flag::NFTokenCreateOffer(
             NFTokenCreateOfferFlag::TfSellOffer,
-        )) && self.amount.get_value_as_u32() == 0
+        )) && amount_as_decimal.is_zero()
         {
             true => Err(NFTokenCreateOfferException::InvalidAmountMustBeGreaterZero),
             false => Ok(()),
```
Comment (sephynox) [inline src/models/transactions/nftoken_create_offer.rs:192]: We should use if/else blocks for booleans.

File: src/models/transactions/nftoken_create_offer.rs [hunk 4]
```diff
@@ -181,9 +183,10 @@ impl<'a> Transaction for NFTokenCreateOffer<'a> {
 
 impl<'a> NFTokenCreateOfferError for NFTokenCreateOffer<'a> {
     fn _get_amount_error(&self) -> Result<(), NFTokenCreateOfferException> {
+        let amount_as_decimal: Decimal = self.amount.clone().try_into().unwrap();
```
Comment (sephynox) [inline src/models/transactions/nftoken_create_offer.rs:186]: This is an unhandled `try_into` and `unwrap`.

**Review by sephynox — CHANGES_REQUESTED** [review]

**Comment by LimpidCrypto** [comment]
Honestly I think it might be better if we had separate models for `XRP` and `IssuedCurrency` as currencies and `XRPAmount` and `IssuedCurrencyAmount` as amounts. That enables us to make our library to be stronger typed. Like an `AccountRoot` objects `Balance` could only be `XRPAmount`.

What's your opinion @sephynox ?
If you agree I would close the PR.

**Pull request closed without merging**